### PR TITLE
Document the batch behaviors of Direct Get

### PR DESCRIPTION
**Pull request by ripienaar**

### CODE REVIEW

File: adr/ADR-31.md [hunk 1]
```diff
@@ -93,19 +98,31 @@ where `last_by_subj` is derived by the token (or series of tokens) following the
 
 It is an error (408) if a client calls Subject-Appended Direct Get and includes a request payload.
 
+#### Batched requests
+
+Using the `batch` and `max_bytes` keys one can request multiple messages in a single API call.
+
+The server will send multiple messages without any flow control to the reply subject, it will send up to `max_bytes` messages.  When `max_bytes` is unset the server will use the Slow Consumer max configuration setting or the server detault (currently 64MB)
```
Comment (derekcollison) [inline adr/ADR-31.md:105]: In server config is `max_pending`

File: adr/ADR-31.md [hunk 2]
```diff
@@ -93,19 +98,31 @@ where `last_by_subj` is derived by the token (or series of tokens) following the
...
+
+The last message in the batch will have the `Nats-Pending-Messages` header set that clients can use to determine if further batch calls are needed. It would also have the `Status` header set to `204` with the `Description` header being `EOB`.
+
 #### Response Format 
 
 Direct Get may return an error code:
+
+- `204` indicates the message is the end of a batch of messages, the description header would have value `EOB`
```
Comment (derekcollison) [inline adr/ADR-31.md:113]: 2xx are not error codes. They are ok w/ conditions.

File: adr/ADR-31.md [hunk 3]
```diff
@@ -93,19 +98,33 @@ where `last_by_subj` is derived by the token (or series of tokens) following the
...
 
+#### Batched requests
+
+Using the `batch` and `max_bytes` keys one can request multiple messages in a single API call.
+
+The server will send multiple messages without any flow control to the reply subject, it will send up to `max_bytes` messages.  When `max_bytes` is unset the server will use the `max_pending` configuration setting or the server default (currently 64MB)
+
+The last message in the batch will have the `Nats-Pending-Messages` header set that clients can use to determine if further batch calls are needed. It would also have the `Status` header set to `204` with the `Description` header being `EOB`.
+
+When requests are made against servers that do not support `batch` the first response will be received and nothing will follow. There is no way to detect this scenario and the batch request will time out. Language documentation must clear call out what server version supports this.
```
Comment (derekcollison) [inline adr/ADR-31.md:109]: clearly call out?

**Review by derekcollison — APPROVED** [review]
LGTM - one nit

File: adr/ADR-31.md (hunk 3 above)
Comment (ripienaar) [reply adr/ADR-31.md]: fixed thanks

**Review by Jarema — APPROVED** [review]
LGTM

**Comment by ripienaar** [comment]
In https://github.com/nats-io/nats-architecture-and-design/pull/262/commits/71b70a7fd6a71a34e0f0a8fdbabac294cb1760a5 I've updated this for what was introduced in https://github.com/nats-io/nats-server/pull/5036

File: adr/ADR-31.md [hunk 4]
```diff
@@ -104,27 +104,29 @@ Using the `batch` and `max_bytes` keys one can request multiple messages in a si
 
 The server will send multiple messages without any flow control to the reply subject, it will send up to `max_bytes` messages.  When `max_bytes` is unset the server will use the `max_pending` configuration setting or the server default (currently 64MB)
 
-The last message in the batch will have the `Nats-Pending-Messages` header set that clients can use to determine if further batch calls are needed. It would also have the `Status` header set to `204` with the `Description` header being `EOB`.
+After the batch is sent a zero length payload message will be sent with the `Nats-Pending-Messages` header set that clients can use to determine if further batch calls are needed. Additionally the `Nats-Last-Sequence` will hold the sequence of the last message sent. It would also have the `Status` header set to `204` with the `Description` header being `EOB`.
```
Comment (derekcollison) [inline adr/ADR-31.md:107]: I think we send in all message now..

File: adr/ADR-31.md [hunk 5]
```diff
@@ -104,27 +104,29 @@ Using the `batch` and `max_bytes` keys one can request multiple messages in a si
...
+Direct Get replies contain the message along with the following message headers:
 
 - `Nats-Stream`: stream name
 - `Nats-Sequence`: message sequence number 
 - `Nats-Time-Stamp`: message publish timestamp
 - `Nats-Subject`: message subject
-- `Nats-Pending-Messages`: the last message in a batch would have this set indicating how many messages are left matching the request
+- `Nats-Num-Pending`: when batched, the number of messages left in the stream matching the batch parameters
+- `Nats-Last-Sequence`: when batched, the stream sequence of the previous message
+- `Nats-Pending-Messages`: the final nil-body message for a batch would have this set indicating how many messages are left matching the request
```
Comment (derekcollison) [inline adr/ADR-31.md:129]: I should make this just NumPending, did not mean to introduce something new AFAIK..

Comment (derekcollison) [reply adr/ADR-31.md:129]: For the messages inside the batch, should the number be relative to the batch? Right now its relative to the whole stream.

Should we do different for inside batch vs the EOB?

Comment (ripienaar) [reply adr/ADR-31.md:129]: Yes, I did find it odd we now have 2 headers - at first I assumed you removed the one but then saw later.  +1 for just using the one.

**Comment by ripienaar** [comment]
Updated for https://github.com/nats-io/nats-server/pull/5069

**Review by derekcollison — APPROVED** [review]
LGTM

**Pull request merged**